pkg/workflow/operation/operation.go

```diff
@@ -269,24 +272,62 @@ func (wo appWorkflowOperator) Rollback(ctx context.Context) error {
 }
 
 // Restart a terminated or finished workflow.
-func (wo appWorkflowOperator) Restart(ctx context.Context) error {
+func (wo appWorkflowOperator) Restart(ctx context.Context, step string) error {
 	app := wo.application
-	if app.Status.Workflow == nil {
+	status := app.Status.Workflow
+	if status == nil {
 		return fmt.Errorf("the workflow in application is not running")
 	}
-	// reset the workflow status to restart the workflow
-	app.Status.Workflow = nil
+	if step == "" {
+		// reset the workflow status to restart the workflow
+		app.Status.Workflow = nil
+
+		if err := wo.cli.Status().Update(ctx, app); err != nil {
+			return err
+		}
 
-	if err := wo.cli.Status().Update(context.TODO(), app); err != nil {
+		return wo.writeOutputF("Successfully restart workflow: %s\n", app.Name)
+	}
+	status.Terminated = false
+	status.Suspend = false
+	status.Finished = false
+	if !status.EndTime.IsZero() {
+		status.EndTime = metav1.Time{}
+	}
+	var cm *corev1.ConfigMap
+	if status.ContextBackend != nil {
+		if err := wo.cli.Get(ctx, client.ObjectKey{Namespace: app.Namespace, Name: status.ContextBackend.Name}, cm); err != nil {
+			return err
+		}
+	}
+	appParser := appfile.NewApplicationParser(wo.cli, nil, nil)
+	appFile, err := appParser.GenerateAppFile(ctx, app)
+	if err != nil {
+		return fmt.Errorf("failed to parse appfile: %w", err)
+	}
+	stepStatus, cm, err := wfUtils.CleanStatusFromStep(appFile.WorkflowSteps, status.Steps, *appFile.WorkflowMode, cm, step)
+	if err != nil {
 		return err
 	}
-
+	status.Steps = stepStatus
+	if err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
```

Comment:
when will the confilct happen? I don't like to use retryOnConflict as it may cover up bugs which can be dangerous.

Reply:
If the controller updates the resources at the same time, it may cause conflict error, it's better to wrap the conflict error for the users so that they don't need to retry.